### DISCUSSION TITLE
Do I need to add checks for the --bind option?

### DISCUSSION THREAD

**Discussion by imlk0** [proot-me/proot-rs#39] [General]
Do we need to add checks to the --bind option of `proot-rs` to only allow binding of dir to dir and file to file?

I bring this up because the behavior of mount --bind is like this :)
![image](https://user-images.githubusercontent.com/14936301/125612121-a0ec939f-ec0b-4d44-94be-ddc203603e30.png)

**Comment by oxr463** [comment]
To clarify, we want to make sure that we only mount files to files and directories to directories?

In the original implementation, this is broken:

```
proot -b file1:dir1
#
proot -b dir1:file1
# 
proot -b file1:file2
#         
proot -b dir1:dir2
#
```

**Reply by imlk0** [reply]
I rechecked and found that the original proot allows all these types of bind above, even non-existent paths as well.

This is quite flexible, so I'm not going to put restrictions on it for now

**Reply by oxr463** [reply]
It is flexible, but is it correct?

The original implementation has a lot of open bugs. I would prefer a more stable tool, if possible.

**Reply by oxr463** [reply]
Test Case / User Question: why would I ever need/want to bind file to directory or any of the other combinations?

**Reply by imlk0** [reply]
> Test Case / User Question: why would I ever need/want to bind file to directory or any of the other combinations?

This provides a way to force an override of the original content to which the path points.

I think this is necessary to provide this capability, just like allowing bind to non-existent paths.

**Reply by imlk0** [reply]
> It is flexible, but is it correct?
> 
> The original implementation has a lot of open bugs. I would prefer a more stable tool, if possible.

Yes, currently there are indeed some problems, for example `readdir(".")` may tell us that `dir1` is a directory, but when we access `./dir1`, we find that it is actually a regular file.

**Reply by imlk0** [reply]
Or we can mark it as an enhancement? Like "binding to a non-existent path"?

**Reply by oxr463** [reply]
Hmm.. I guess that makes sense.

It doesn't need to be an enhancement, but we should definitely document the behavior.
